Describe the outbox pattern (#4257)
MicrosoftDocs/architecture-center: docs/patterns/outbox.md

While this characteristic is great for ensuring no information is lost by the running system, it has an inherent drawback that any given message may be delivered to the processing code multiple times. In absence of code that guards against processing a message multiple times the data quickly becomes corrupted by updates applied multiple times.

Historically this issue has been mitigated by distributed transcation technologies implementing Two-Phase Commit protocol (2PC) like Microsoft Distributed Transaction Coordinator. Such technologies allowed a single transaction to span both the messaging infrastructure and the database, ensuring that either:
 - all the data is updated and the message is consumed
 - no data is updated and the message is returned back to the queue

Due to the issues with scalability and performance of the 2PC protocol, the distributed transaction technologies are no longer available for modern messaging technologies, such as Azure Service Bus and developers need to write code protecting against duplicate processing into their message handling logic. Adding such logic to a message handling code is referred to as making that code _idempotent_. 

> In mathematics, where the term originates, _idempotence_ is a property of fuctions. A given function _f_ is _idempotent_ if and only if _f(f(x)) = (x)_. In the cotext of the HTTP protocol, [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4), states, "A ... method is considered _idempotent_ if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request.". The latter is hepful in attempt to define _idempotence_ for the message handling problem:

Reviewer: ```suggestion
In mathematics, where the term originates, _idempotence_ is a property of functions. A given function _f_ is _idempotent_ if and only if _f(f(x)) = f(x)_. In the context of the HTTP protocol, [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4), states, "A ... method is considered _idempotent_ if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request.". The latter is helpful in attempt to define _idempotence_ for the message handling problem:
```

> An independent message handling code is one that produces the same side effects regardless of how many times a given message has been delivered

> The well-documented approach to the challenge of making message handling code _idempotent_ is to add a check, prior to executing the code, if the side effects have already been applied. This is illustrated by the following pseudocode:

Reviewer: ```suggestion
The well-documented approach to the challenge of making message handling code _idempotent_ is to add a check, prior to executing the code, if the code was already executed, usually by checking some external resource like a database. This is illustrated by the following pseudocode:
```

```c#

> if (!HasBeenProcessed(message))

Reviewer: ```suggestion
if (HasBeenProcessed(message))
```

{
    return;
}
Process(message);
MarkProcessed(message);
```

> It does not, however, take into account that a typical message handling code affects multiple external resources, i.e. a database (store or update some data) and a messaging infrastructure (send follow-up messages). In other words, it may happen that the first attempt at processing a message resulted in updating a database but failed to send out messages. The message is put back into the queue, picked up again and the second attempt manages to send the messages. 

Reviewer: ```suggestion
It does not, however, take into account that typical message handling code affects multiple external resources, i.e. a database (store or update some data) and messaging infrastructure (send follow-up messages). In other words, it may happen that the first attempt at processing a message resulted in updating the database but crashed before being able to send out messages. In that case, the message is "rolled back" to the queue, after which it will be picked up and attempted to be processed again. 
```

> In order for that logic to be consistent, both attempts have to see exactly the same state, including but not limited to:

Reviewer: ```suggestion
In order for the system to arrive at an eventually consistent state, the logic processing the message the second time must correctly emit the messages that should have been emitted after the first processing attempt, but not make any other changes to the database. This can be challenging when the message processing logic incorporates thing like:
```

 - data in all queried data stores
 - environment properties, e.g. data and time
 - pseudo-random value generators

>  - sequeces, e.g. GUID

Reviewer: ```suggestion
 - sequences, e.g. GUID
```

## Solution

The solution to this problem is to split the execution of the message handling logic into two parts:

>  - compute the side effects

Reviewer: I'm not sure it is good to use a term like "side effects". I think some developers would not intuitively understand the intent.

 - apply the side effects

and is based on the assumption that, once side effects are computed and made persistent, they can always be applied. This, of course, is not true for most interactions with database in any environment that allows concurret access. For that reason, one resource, usually the database, is designated as primary. The primary resource plays double role. In addition to being the target for the side effects, it also is used to store the computed side effects that target other resources. The resulting algorithm for the Outbox pattern is following:

```c#

> if (!HasBeenProcessed(message))

Reviewer: ```suggestion
if (HasBeenProcessed(message))
```

{
    return;
}
ComputeSideEffects(message);
var tx = BeginTransaction();
ApplySideEffectsToPrimary(tx);
StoreSideEffects(tx);
MarkProcessed(message, tx);
tx.Commit();
CleanUpSideEffects(tx);
```

The Outbox pattern can be implemented in the infrastructure layer of the solution ensuring that all message handlers running on top of that infrastructure are automatically made _idempotent_. The alternative is ensure each individual message handler is _idempotent_ by applying techniques some of the following techniques:

- do not access more than one resource in any message handler
- use deterministic or domain identifiers instead of GUIDs
- do not rely on system clock or pseudo-random value generators
- use idempotent-by-design data structures

### Benefits

> - Provides programming simple programming abstraction, simialar to one exposed by distributed transaction technologies

Reviewer: ```suggestion
- Provides a simple programming abstraction, similar to one exposed by distributed transaction technologies
```

- Can be implemented in the infrastructure layer with minimal effect to message handling code, allowing smooth transition away from distributed transaction technologies
- Supports resources that never had been able to participate in distributed transactions, such as key-value stores
- Has relatively mild requirements for the primary resource -- optimistic concurrency control

### Drawbacks

- Generates more traffic with the primary resource (database) in form of two additional operations requiring a round-trip: checking if a message has been processed and cleaning up the side effects.
- Requires that the primary resource supports optimistic concurrency control
- Requires storing information about all previously processed messages. In principle that information should never be removed but in practice often age-based approach is used to prevent data from growing indefinitely. 

## Issues and considerations

Consider the following points when implementing the Outbox pattern:

> - In order to take advatage of the pattern, each message needs to have a unique identifier. Most messaging technologies, e.g. Azure Service Bus, provide this identifier out-of-the-box, but all.

Reviewer: ```suggestion
- In order to take advantage of the pattern, each message needs to have a unique identifier. Many messaging technologies like Azure Service Bus provide this identifier out-of-the-box, but in some of them, it is not set automatically.
```

StoreSideEffects(tx);
MarkProcessed(message, tx);
tx.Commit();

ApplySideEffects();
CleanUpSideEffects();
```

The Outbox pattern can be implemented in the infrastructure layer of the solution ensuring that all message handlers running on top of that infrastructure are automatically made _idempotent_. The alternative is ensure each individual message handler is _idempotent_ by applying techniques some of the following techniques:

- do not access more than one resource in any message handler
- use deterministic or domain identifiers instead of GUIDs
- do not rely on system clock or pseudo-random value generators
- use idempotent-by-design data structures

### Benefits

- Provides a simple programming abstraction, similar to one exposed by distributed transaction technologies
- Can be implemented in the infrastructure layer with minimal effect to message handling code, allowing smooth transition away from distributed transaction technologies
- Supports resources that never had been able to participate in distributed transactions, such as key-value stores
- Has relatively mild requirements for the primary resource -- optimistic concurrency control

### Drawbacks

> - Generates more traffic with the primary resource (database) in form of two additional operations requiring a round-trip: checking if a message has been processed and cleaning up the side effects.

Reviewer 2: ```suggestion
- Puts more load on the primary resource (database)- two additional operations per message
- Increases message processing delay- checking if a message has been processed and cleaning up the side effects both required database round-trip
```

- Requires that the primary resource supports optimistic concurrency control
- Requires storing information about all previously processed messages. In principle that information should never be removed but in practice often age-based approach is used to prevent data from growing indefinitely. 

## Issues and considerations

Consider the following points when implementing the Outbox pattern:

> - In order to take advantage of the pattern, each message needs to have a unique identifier. Many messaging technologies like Azure Service Bus provide this identifier out-of-the-box, but in some of them, it is not set automatically.

Reviewer 2: ```suggestion
- In order to take advantage of the pattern, each message needs to have a unique identifier. Many messaging technologies like Azure Service Bus provide this identifier out-of-the-box, but in some cases, manual handling is needed.
```

> - Consider how long to store information about processed messages. The likelihood of message being duplicated diminishes with time significantly. In practice value of 1 or 2 weeks is often used as maximum age to keep the deduplication data.

Reviewer 2: ```suggestion
- The amount of time for which processed messages should be stored depends on the concrete scenario. For most of the systems, the likelihood of a message being duplicated diminishes significantly with time. In practice value of 1 or 2 weeks is often used as the maximum age for keeping the deduplication data.
```

> - Depending on the technology used for the primary resource, it might or might not be possible to use native features to remove expired dedupliction data. For example, Cosmos DB has a built in time-to-live mechanism that can be used to ensure message processing information is removed after configured time.

Reviewer 2: ```suggestion
- Depending on the technology used for the primary resource, it might or might not be possible to use native features to remove expired dedupliction data. Cosmos DB users can take advantage of the built-in time-to-live mechanism to ensure message processing information is removed after the configured time.
```

> - The clock for removing deduplication data can only be started after the `CleanUpSideEffects` operation succeeds. Otherwise there is a risk of losing the side effects if a message is stuck for extended period of time, e.g. falsly identified as a _poison message_ and sent to a _dead-letter queue_.

Reviewer 2: ```suggestion
- The countdown process for removing deduplication data can only be started after the `CleanUpSideEffects` operation succeeds. Otherwise, the side effects data can be prematurely removed for messages "stuck" for an extended time, e.g. falsely identified as a _poison message_ and sent to a _dead-letter queue_.
```

- Side effects other than sending or publishing messages are supported provided the likelihood of failure to apply them is negligible, e.g. creating or updating a document in a key-value store.

### Side effects application

The design of the side effects application is a major consideration when implementing the outbox pattern. There are three major approaches to the problem.

#### Background worker

> In this approach an external process periodically polls the data store where the side effects are persisted. The advantage of this approach is the fact that it can also be used at the boundary of the system e.g. when processing incoming HTTP requests and sending out messagess. There are, however, significant disadvantages of this approach. First, it relies on querying the data store for all outstanding side effects. This query pattern might be expensive or even impossible in certain partitioned data stores (e.g. Cosmos DB). The second issue is the fact that polling implies the code is executed periodically. Such requirement makes deployment in serverless environments (e.g. Azure Functions) problematic as the worker needs to use a dedicated (timer) trigger. Finally, the polling mechanism has poor scalability characteristics. Higher side effect processing throughput cannot be simply achieved by deployment multiple background workers as they would compete to process the same items. 

Reviewer 2: ```suggestion
In this approach, an external process periodically polls the data store where the side effects are persisted. The advantage of this approach is the fact that it can also be used at the boundary of the system e.g. when processing incoming HTTP requests and sending out messages. There are, however, significant disadvantages. 

First, it relies on querying the data store for all outstanding side effects. This query pattern might be expensive or even impossible in certain partitioned data stores (e.g. Cosmos DB). Secondly, polling requires the code to be executed periodically (at intervals). Such a requirement makes deployment in serverless environments (e.g. Azure Functions) problematic as the worker needs a dedicated (timer) trigger. Finally, the polling mechanism has poor scalability characteristics. Higher side effect processing throughput cannot be simply achieved by deploying multiple background workers as they would compete to process the same items. 
```

#### Incoming message

An alternative approach is to use the incoming message as a driver for side effects application. In this approach the side effects are applied immediately after the main transaction is committed and before the incoming message is consumed. If the application fails, the incoming message simply goes back to the queue and is picked up again. This time the business processing is skipped and only the side effects application is performed.

> This approach scales well, does not require any dedicated deployment items and does not rely on cross-partition queries. Its disadvantage is the fact that it cannot be used at the system boundary.

Reviewer 2: ```suggestion
This approach scales well, does not require any dedicated deployment items and does not rely on cross-partition queries. Its disadvantage is the fact that it cannot be used at the system boundary when no incoming message exists.
```

#### Control message

In this approach the side effects application is driven by a special control message that is dispatched prior to committing the main transaction. The processing logic of this message has a built-in delay mechanism to give the main transaction some time to commit (e.g. 30 seconds). If that does not happen, a special record is created that prevents the transaction completion and the control message is dropped. 

> This method has the same benefits as the approach based on the incoming message but has been adapted for usage at the boundary of the system where there is not incoming durable message.

Reviewer 2: ```suggestion
This method has the same benefits as the approach based on the incoming message but can be used at the boundary of the system at the cost of additional control message.
```

## When to use this pattern

Use the Outbox pattern when you:

> - Use modern messaging technology, such as Azure Service Bus, to build line-of-business applictions where you cannot afford to lose messages or duplicate the processing. 

Reviewer 2: ```suggestion
- When using modern messaging technology, such as Azure Service Bus, to build line-of-business applications where you cannot afford to lose messages or duplicate the processing. 
```
